### ISSUE TITLE
Parameters to fn:sort

### ISSUE THREAD

**Issue by michaelhkay** [qt4cg/qtspecs#1085]
An interesting suggestion made in passing in the thread discussing fn:ranks(). It would be possible to combine the `collation` argument and the `ascending/descending` argument of `fn:sort` into a single argument, whose value is an optional "ascending|descending" keyword followed by an optional collation URI (whitespace-separated, presumably).

This might seem a little bizarre at first sight, but having a list of collation URIs followed by a list of sort key functions followed by a list of ascending/descending keywords is also a little bizarre, and it would have two advantages - it would make better use of the second argument which is currently nearly always set to `()`, and it would put the two parts of the order specification (the collation and its direction) in closer proximity. After all, they are used in combination to decide whether one value precedes or follows another.

I'm not 100% convinced by the idea, but it seems worth considering. What do people think?

**Comment by ChristianGruen** [comment]
I would call it a pragmatic hack… I’m torn, as it would only be useful for the secondary cases of descending sorts. For ascending sorts, `sort($input, (), $fn)` would still be shorter, and it’s already familiar to users of XPath 3.1.

How would the following query need to be written with the proposed syntax?

```xquery
sort(
  $persons,
  keys   := (fn { @age }, fn { @name }),
  orders := ('ascending', 'descending')
)
```

In New-Language-for-XML-and-JSON-data 1.0, I would fully comply with Dimitre that `keys` should be placed as second argument.

**Comment by michaelhkay** [comment]
That call on sort could be written

```
sort($persons, ('ascending',  'descending'),
                          (fn { @age },  fn { @name }))
```

With a collation, it might become

```
sort($persons, ('ascending',  'descending http://www.w3.org/2013/collation/UCA?lang=SE'),
                          (fn { @age },  fn { @name }))
```

**Comment by dnovatchev** [comment]
> That call on sort could be written
> 
> ```
> sort($persons, ('ascending',  'descending'),
>           (fn { @age },  fn { @name }))
> ```
> 
> With a collation, it might become
> 
> ```
> sort($persons, ('ascending',  'descending http://www.w3.org/2013/collation/UCA?lang=SE'),
>              (fn { @age },  fn { @name }))
> ```

Even more compact:

```xq
sort($persons, 
       fn($age, $name, 
         map{'paramName': 'name', 
             'collation': 'http://www.w3.org/2013/collation/UCA?lang=SE',
             'order': 'descending'} )
   {$age || '*' || $name}
```

This specifies a single **`key()`** function that takes as parameters the properties (sub-parts) of each of the input items, on which the sort must be done, and a sequence of maps - one for each of these properties, who are either strings with non-default collation, or reverse sorting order, or numeric with reverse sorting order.

Thus everything that may be omitted (would use the default order, or collation) is really omitted. 

This is just a quick first attempt. Maybe this can be further optimized by always passing only a single map containing the order-collation details in submaps that are values for a key whose value is the name of the corresponding parameter (and included in the map only if details other than the defaults are needed for this parameter).

**Comment by michaelhkay** [comment]
```
sort($persons, 
       fn($age, $name, 
         map{
```
The whole point of the current 4.0 spec is to retain backwards compatibility with the 3.1 interface.

**Comment by dnovatchev** [comment]
> ```
> sort($persons, 
>        fn($age, $name, 
>          map{
> ```
> 
> The whole point of the current 4.0 spec is to retain backwards compatibility with the 3.1 interface.

Yes, and for new functions we can and should provide better user experience.

This is just a hint of what is possible. History teaches us that even constitutions of nations often have series of amendments (such as the abolition of slavery, or granting the women the right to vote, ..., etc).

It is instructive how we are trying to extend the functionality of existing functions, by trying not to affect backwards compatibility. 

In  many other programming languages we can observe that a new function or datatype is introduced that extends the currently existing function/datatype. One example is the T-SQL **_[datetime2](https://learn.microsoft.com/en-us/sql/t-sql/data-types/datetime2-transact-sql?view=sql-server-ver16)_** datatype.

Another example again from SQL Server: https://learn.microsoft.com/en-us/sql/t-sql/data-types/ntext-text-and-image-transact-sql?view=sql-server-ver16

"**_The ntext, text, and image data types will be removed in a future version of SQL Server. Avoid using these data types in new development work, and plan to modify applications that currently use them. Use [nvarchar(max)](https://learn.microsoft.com/en-us/sql/t-sql/data-types/nchar-and-nvarchar-transact-sql?view=sql-server-ver16), [varchar(max)](https://learn.microsoft.com/en-us/sql/t-sql/data-types/char-and-varchar-transact-sql?view=sql-server-ver16), and [varbinary(max)](https://learn.microsoft.com/en-us/sql/t-sql/data-types/binary-and-varbinary-transact-sql?view=sql-server-ver16) instead._**"

So, in new functions I would totally support having better signatures that enhance the user experience.
As a user I will not be surprised to see functions named in a similar way, such as `fn:{oldName}2`, or using any other preferred convention.

**Comment by ChristianGruen** [comment]
I propose to close the issue. I am not sure if the proposed change would make it much easier to use the function.

**Comment by michaelhkay** [comment]
Dropped the label, I would like to have a stab at solving this one.

**Comment by michaelhkay** [comment]
I'd like to suggest that we retain the 3.1 function as defined in 3.1, and put the 4.0 functionality in a new function

`fn:sort-by($input as item()*, $keys as record(key as fn($item), collation? as xs:string, order? as enum('ascending', 'descending'))*)`

A simple call might be 

`sort-by(//employee, {'key': fn{surname}})`

With a collation:

`sort-by(//employee, {'key': fn{surname}, 'collation': collation('lang':'de')}})`

With two keys:

```
sort-by(//employee, (
          {'key': fn{surname}}, 
          {'key': fn{number(salary)}, 'order':'descending'} ))
```
